lib/ansible/plugins/action/reboot.py

```diff
@@ -94,18 +94,31 @@ def construct_command(self):
         return reboot_command
 
     def get_system_boot_time(self):
+        stdout = ''
+        stderr = ''
         command_result = self._low_level_execute_command(self.DEFAULT_BOOT_TIME_COMMAND, sudoable=self.DEFAULT_SUDOABLE)
 
         # For single board computers, e.g., Raspberry Pi, that lack a real time clock and are using fake-hwclock
         # launched by systemd, the update of utmp/wtmp is not done correctly.
         # Fall back to using uptime -s for those systems.
         # https://github.com/systemd/systemd/issues/6057
         if '1970-01-01 00:00' in command_result['stdout']:
+            stdout += command_result['stdout']
+            stderr += command_result['stderr']
             command_result = self._low_level_execute_command('uptime -s', sudoable=self.DEFAULT_SUDOABLE)
 
+        # This is a last resort for bare Linux systems (e.g. OpenELEC) where 'who -b' or 'uptime -s' are not supported.
+        # Other options like parsing /proc/uptime or default uptime output are less reliable than this
         if command_result['rc'] != 0:
```

Comment:
One issue with this is that `win_reboot` uses this same method. So if the windows `DEFAULT_BOOT_TIME_COMMAND` results in an rc != 0, it will then try this `cat` command, which will gloriously fail.

The previous conditional which runs `uptime -s` won't ever match on Windows because that string will never match the output of `(Get-WmiObject -ClassName Win32_OperatingSystem).LastBootUpTime`.

This is the point where it would be helpful to have platform/dist as a class attribute, either as an additional check here, or to look up the correct boot time command based on platform/dist. I'm also open to other ideas.

Reply:
In my WIP PR, I did change the "get the last boot time" command to be a variable, so that will remove the hard coding here.

Will fix in future™.

Still, I can see it being confusing to someone running `win_reboot` (assuming the first command failed to get the last boot time) if they get this error:

```
win_reboot: failed to get host boot time info, rc: 1, stdout: 20180913204044.576906+000
, stderr: cat : Cannot find path 'C:\proc\sys\kernel\random\boot_id' because it does not exist.
At line:1 char:1
+ cat /proc/sys/kernel/random/boot_id
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (C:\proc\sys\kernel\random\boot_id:String) [Get-Content], ItemNotFoundEx
ception
+ FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetContentCommand
```

Evidently Windows has a `cat` command now!?

Reply:
Well, the distinction between Windows and Unix is now easy because it's a different entry-point. So we don't need to do a `cat` on Windows. That was my main point. But I also don't mind having the fallback mechanism on Unix, that first does `who -b`, then `uptime -s`, and then `cat /.../boot_id`. Which to me is quite sensible as well. It's a trade-off for simplicity and in most cases it only requires a single command `who -b`. If you implement support for getting the platform, you will *always* have to run at least 2 commands.

So basically what I am saying is that in general the fallback mechanism is more efficient than targeting based on platform.

Reply:
And yes, PowerShell has cmdlets for common Unix commands, which is freaking a lot of people out (because they are not drop-in replacements). https://daniel.haxx.se/blog/2016/08/19/removing-the-powershell-curl-alias/